README.pod
=pod

=encoding utf-8

=head1 NAME

MongoDBx::Queue - A message queue implemented with MongoDB

=head1 VERSION

version 0.004

=head1 SYNOPSIS

  use v5.10;
  use MongoDB;
  use MongoDBx::Queue;

  my $connection = MongoDB::Connection->new( @parameters );
  my $database = $connection->get_database("queue_db");

  my $queue = MongoDBx::Queue->new( { db => $database } );

  $queue->add_task( { msg => "Hello World" } );
  $queue->add_task( { msg => "Goodbye World" } );

  while ( my $task = $queue->reserve_task ) {
    say $task->{msg};
    $queue->remove_task( $task );
  }

=head1 DESCRIPTION

B<ALPHA> -- this is an early release and is still in development.  Testing
and feedback welcome.

MongoDBx::Queue implements a simple, prioritized message queue using MongoDB as
a backend.  By default, messages are prioritized by insertion time, creating a
FIFO queue.

On a single host with MongoDB, it provides a zero-configuration message service
across local applications.  Alternatively, it can use a MongoDB database
cluster that provides replication and fail-over for an even more durable,
multi-host message queue.

Features:

=over 4

=item *

messages as hash references, not objects

=item *

arbitrary message fields

=item *

arbitrary scheduling on insertion

=item *

atomic message reservation

=item *

stalled reservations can be timed-out

=item *

task rescheduling

=back

Not yet implemented:

=over 4

=item *

parameter checking

=item *

error handling

=back

Warning: do not use with capped collections, as the queued messages will not
meet the constraints required by a capped collection.

=head1 ATTRIBUTES

=head2 db

A MongoDB::Database object to hold the queue.  Required.

=head2 name

A collection name for the queue.  Defaults to 'queue'.  The collection must
only be used by MongoDBx::Queue or unpredictable awful things will happen.

=head2 safe

Boolean that controls whether 'safe' inserts/updates are done.
Defaults to true.

=head1 METHODS

=head2 new

  $queue = MongoDBx::Queue->new( { db => $database, %options } );

Creates and returns a new queue object.  The C<db> argument is required.
Other attributes may be provided as well.

=head2 add_task

  $queue->add_task( \%message, \%options );

Adds a task to the queue.  The C<\%message> hash reference will be shallow
copied into the task and not include objects except as described by
L<MongoDB::DataTypes>.  Top-level keys must not start with underscores, which are
reserved for MongoDBx::Queue.

The C<\%options> hash reference is optional and may contain the following key:

=over 4

=item *

C<priority>: sets the priority for the task. Defaults to C<time()>.

=back

Note that setting a "future" priority may cause a task to be invisible
to C<reserve_task>.  See that method for more details.

=head2 reserve_task

  $task = $queue->reserve_task;
  $task = $queue->reserve_task( \%options );

Atomically marks and returns a task.  The task is marked in the queue as
"reserved" (in-progress) so it can not be reserved again unless is is
rescheduled or timed-out.  The task returned is a hash reference containing the
data added in C<add_task>, including private keys for use by MongoDBx::Queue
methods.

Tasks are returned in priority order from lowest to highest.  If multiple tasks
have identical, lowest priorities, their ordering is undefined.  If no tasks
are available or visible, it will return C<undef>.

The C<\%options> hash reference is optional and may contain the following key:

=over 4

=item *

C<max_priority>: sets the maximum priority for the task. Defaults to C<time()>.

=back

The C<max_priority> option controls whether "future" tasks are visible.  If
the lowest task priority is greater than the C<max_priority>, this method
returns C<undef>.

=head2 reschedule_task

  $queue->reschedule_task( $task );
  $queue->reschedule_task( $task, \%options );

Releases the reservation on a task so it can be reserved again.

The C<\%options> hash reference is optional and may contain the following key:

=over 4

=item *

C<priority>: sets the priority for the task. Defaults to the task's original priority.

=back

Note that setting a "future" priority may cause a task to be invisible
to C<reserve_task>.  See that method for more details.

=head2 remove_task

  $queue->remove_task( $task );

Removes a task from the queue (i.e. indicating the task has been processed).

=head2 apply_timeout

  $queue->apply_timeout( $seconds );

Removes reservations that occurred more than C<$seconds> ago.  If no
argument is given, the timeout defaults to 120 seconds.  The timeout
should be set longer than the expected task processing time, so that
only dead/hung tasks are returned to the active queue.

=head2 search

  my @results = $queue->search( \%query, \%options );

Returns a list of tasks in the queue based on search criteria.  The
query should be expressed in the usual MongoDB fashion.  In addition
to MongoDB options C<limit>, C<skip> and C<sort>, this method supports
a C<reserved> option.  If present, results will be limited to reserved
tasks if true or unreserved tasks if false.

=head2 peek

  $task = $queue->peek( $task );

Retrieves a full copy of the task from the queue.  This is useful to retrieve all
fields from a partial-field result from C<search>.  It is equivalent to:

  $self->search( { _id => $task->{_id} } );

Returns undef if the task is not found.

=head2 size

  $queue->size;

Returns the number of tasks in the queue, including in-progress ones.

=head2 waiting

  $queue->waiting;

Returns the number of tasks in the queue that have not been reserved.

=for :stopwords cpan testmatrix url annocpan anno bugtracker rt cpants kwalitee diff irc mailto metadata placeholders metacpan

=head1 SUPPORT

=head2 Bugs / Feature Requests

Please report any bugs or feature requests through the issue tracker
at L<https://github.com/dagolden/mongodbx-queue/issues>.
You will be notified automatically of any progress on your issue.

=head2 Source Code

This is open source software.  The code repository is available for
public review and contribution under the terms of the license.

L<https://github.com/dagolden/mongodbx-queue>

  git clone git://github.com/dagolden/mongodbx-queue.git

=head1 AUTHOR

David Golden <dagolden@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2012 by David Golden.

This is free software, licensed under:

  The Apache License, Version 2.0, January 2004



Changes
Revision history for MongoDBx-Queue

{{$NEXT}}

2.002     2020-04-07 17:16:42-04:00 America/New_York

    [CHANGED]

    - As an optimization, the 'size' method now gets a faster, estimated count.

2.001     2019-06-30 23:26:26-04:00 America/New_York

    [ADDED]

    - Added 'version' option to toggle between legacy 'v1' implementation and
      a new 'v2' implementation with better index coverage.

2.000     2018-10-04 15:44:27-04:00 America/New_York

    [INCOMPATIBLE CHANGES]

    - Requires MongoDB driver version 2.0.0 and MooseX::Role::MongoDB
      v1.0.0

    - 'safe' option removed; all reads and writes now use 'majority' read
      and write concerns

1.002     2014-01-24 23:05:20-05:00 America/New_York

    [FIXES]

    - Updated for changes in MooseX::Role::MongoDB 0.006

1.001     2013-10-15 19:35:26 America/New_York

    [FIXES]

    - Consumes MooseX::Role::Logger as required by MooseX::Role::MongoDB
      version 0.003 or later

1.000     2013-09-22 20:57:32 America/New_York

    [INCOMPATIBLE CHANGES]

    - MongoDBx::Queue prior to version 1.000 were not fork-safe.  A fork
      could cause MongoDB connections to hang.  As of 1.000, MongoDB
      connections are managed via MooseX::Role::MongoDB, which provides
      fork-safety.

    - As result of this change, attributes have changed.  MongoDBx::Queue
      no longer has a C<db> attributed taking a MongoDB::Database argument
      and the C<name> attribute has been renamed to C<collection_name>

    - Now, you provide C<client_options> controlling the connection to
      the database.  You provide C<database_name> to indicate the database
      to hold your queue.  And you provide C<collection_name> to indicate
      the collection to hold the queue.

    - No attributes are required any more.  By default, MongoDBx::Queue
      will connect to localhost on the default MongoDB port, will use
      the 'test' database and the 'queue' collection.

    [ADDED]

    - Automatically ensures an index on priority is in place on the
      collection

    [FIXES]

    - Fixed task reservation on Perl v5.18 that borke due to hash-order
      randomization. MongoDB expects command documents to be ordered a
      certain way so we now use array references instead of hash references
      to pass such commands.

    [INTERNAL]

    - Dropped Const::Fast as prerequisite

0.004     2013-04-17 20:06:53 America/New_York

    [FIXED]

    - Have peek() do the right thing for list or scalar return context

0.003     2013-03-28 18:34:49 America/New_York

    [ADDED]

    - Added 'search' and 'peek' methods to allow inspection of the queue

0.002     2012-09-05 18:08:14 America/New_York

    [INCOMPATIBLE CHANGES]

    - reschedule_task() now takes a hashref of options using the same format
        as add_task() instead of taking a scalar epoch seconds time argument

    [ADDED]

    - add_task() now supports setting arbitrary priority

    [CHANGED]

    - reserve_task() only find tasks up to a maximum priority; defaults to
        time(); this makes "future" tasks invisible unless a different
        maximum priority is requested

    - generally improved documentation for consistency describing new
        features

0.001     2012-08-30 17:34:16 America/New_York

    - First release

# vim: ts=4 sts=4 sw=4 et:


lib/MongoDBx/Queue.pm
use 5.010;
use strict;
use warnings;

package MongoDBx::Queue;

# ABSTRACT: A message queue implemented with MongoDB

our $VERSION = '2.003';

use Moose 2;
use MooseX::Types::Moose qw/:all/;
use MooseX::AttributeShortcuts;

use MongoDB 2 ();
use namespace::autoclean;

with (
    'MongoDBx::Queue::Role::_CommonOptions',
);

#--------------------------------------------------------------------------#
# Public attributes
#--------------------------------------------------------------------------#

=attr database_name

A MongoDB database name.  Unless a C<db_name> is provided in the
C<client_options> attribute, this database will be the default for
authentication.  Defaults to 'test'

=attr client_options

A hash reference of L<MongoDB::MongoClient> options that will be passed to its
C<connect> method.

=attr collection_name

A collection name for the queue.  Defaults to 'queue'.  The collection must
only be used by MongoDBx::Queue or unpredictable awful things will happen.

=attr version

The implementation version to use as a backend.  Defaults to '1', which is the
legacy implementation for backwards compatibility.  Version '2' has better
index coverage and will perform better for very large queues.

B<WARNING> Versions are not compatible.  You MUST NOT have V1 and V2 clients
using the same database+collection name.  See L</MIGRATION BETWEEN VERSIONS>
for more.

=cut

has version => (
    is      => 'ro',
    isa     => Int,
    default => 1,
);

#--------------------------------------------------------------------------#
# Private attributes and builders
#--------------------------------------------------------------------------#

has _implementation => (
    is => 'lazy',
    handles => [ qw(
        add_task
        reserve_task
        reschedule_task
        remove_task
        apply_timeout
        search
        peek
        size
        waiting
    )],
);

sub _build__implementation {
    my ($self) = @_;
    my $options = {
        client_options => $self->client_options,
        database_name => $self->database_name,
        collection_name => $self->collection_name,
    };
    if ($self->version == 1) {
        require MongoDBx::Queue::_V1;
        return MongoDBx::Queue::_V1->new($options);
    }
    elsif ($self->version == 2) {
        require MongoDBx::Queue::_V2;
        return MongoDBx::Queue::_V2->new($options);
    }
    else {
        die "Invalid MongoDBx::Queue 'version' (must be 1 or 2)"
    }
}

sub BUILD {
    my ($self) = @_;
    $self->_implementation->create_indexes;
}

#--------------------------------------------------------------------------#
# Public method documentation
#--------------------------------------------------------------------------#

=method new

   $queue = MongoDBx::Queue->new(
        version => 2,
        database_name   => "my_app",
        client_options  => {
            host => "mongodb://example.net:27017",
            username => "willywonka",
            password => "ilovechocolate",
        },
   );

Creates and returns a new queue object.

=method add_task

  $queue->add_task( \%message, \%options );

Adds a task to the queue.  The C<\%message> hash reference will be shallow
copied into the task and not include objects except as described by
L<MongoDB::DataTypes>.  Top-level keys must not start with underscores, which are
reserved for MongoDBx::Queue.

The C<\%options> hash reference is optional and may contain the following key:

=for :list
* C<priority>: sets the priority for the task. Defaults to C<time()>.

Note that setting a "future" priority may cause a task to be invisible
to C<reserve_task>.  See that method for more details.

=method reserve_task

  $task = $queue->reserve_task;
  $task = $queue->reserve_task( \%options );

Atomically marks and returns a task.  The task is marked in the queue as
"reserved" (in-progress) so it can not be reserved again unless is is
rescheduled or timed-out.  The task returned is a hash reference containing the
data added in C<add_task>, including private keys for use by MongoDBx::Queue
methods.

Tasks are returned in priority order from lowest to highest.  If multiple tasks
have identical, lowest priorities, their ordering is undefined.  If no tasks
are available or visible, it will return C<undef>.

The C<\%options> hash reference is optional and may contain the following key:

=for :list
* C<max_priority>: sets the maximum priority for the task. Defaults to C<time()>.

The C<max_priority> option controls whether "future" tasks are visible.  If
the lowest task priority is greater than the C<max_priority>, this method
returns C<undef>.

=method reschedule_task

  $queue->reschedule_task( $task );
  $queue->reschedule_task( $task, \%options );

Releases the reservation on a task so it can be reserved again.

The C<\%options> hash reference is optional and may contain the following key:

=for :list
* C<priority>: sets the priority for the task. Defaults to the task's original priority.

Note that setting a "future" priority may cause a task to be invisible
to C<reserve_task>.  See that method for more details.

=method remove_task

  $queue->remove_task( $task );

Removes a task from the queue (i.e. indicating the task has been processed).

=method apply_timeout

  $queue->apply_timeout( $seconds );

Removes reservations that occurred more than C<$seconds> ago.  If no
argument is given, the timeout defaults to 120 seconds.  The timeout
should be set longer than the expected task processing time, so that
only dead/hung tasks are returned to the active queue.

=method search

  my @results = $queue->search( \%query, \%options );

Returns a list of tasks in the queue based on search criteria.  The
query should be expressed in the usual MongoDB fashion.  In addition
to MongoDB options (e.g. C<limit>, C<skip> and C<sort>) as described
in the MongoDB documentation for L<MongoDB::Collection/find>, this method
supports a C<reserved> option.  If present, results will be limited to reserved
tasks if true or unreserved tasks if false.

=method peek

  $task = $queue->peek( $task );

Retrieves a full copy of the task from the queue.  This is useful to retrieve all
fields from a projected result from C<search>.  It is equivalent to:

  $self->search( { _id => $task->{_id} } );

Returns undef if the task is not found.

=method size

  $queue->size;

Returns the number of tasks in the queue, including in-progress ones.

=method waiting

  $queue->waiting;

Returns the number of tasks in the queue that have not been reserved.

=cut

__PACKAGE__->meta->make_immutable;

1;

=for Pod::Coverage BUILD

=head1 SYNOPSIS

    use v5.10;
    use MongoDBx::Queue;

    my $queue = MongoDBx::Queue->new(
        version => 2,
        database_name => "queue_db",
        client_options => {
            host => "mongodb://example.net:27017",
            username => "willywonka",
            password => "ilovechocolate",
        }
    );

    $queue->add_task( { msg => "Hello World" } );
    $queue->add_task( { msg => "Goodbye World" } );

    while ( my $task = $queue->reserve_task ) {
        say $task->{msg};
        $queue->remove_task( $task );
    }

=head1 DESCRIPTION

MongoDBx::Queue implements a simple, prioritized message queue using MongoDB as
a backend.  By default, messages are prioritized by insertion time, creating a
FIFO queue.

On a single host with MongoDB, it provides a zero-configuration message service
across local applications.  Alternatively, it can use a MongoDB database
cluster that provides replication and fail-over for an even more durable,
multi-host message queue.

Features:

=for :list
* messages as hash references, not objects
* arbitrary message fields
* arbitrary scheduling on insertion
* atomic message reservation
* stalled reservations can be timed-out
* task rescheduling
* automatically creates correct index
* fork-safe

Not yet implemented:

=for :list
* parameter checking
* error handling

Warning: do not use with capped collections, as the queued messages will not
meet the constraints required by a capped collection.

=head1 MIGRATION BETWEEN VERSIONS

Implementation versions are not compatible.  Migration of active tasks from
version '1' to version '2' is an exercise left to end users.

One approach to migration could be to run a script with two C<MongoDBx::Queue>
clients, one using version '1' and one using version '2', using different
C<database_name> attributes.  Such a script could iteratively reserve a task
with the v1 client, add the task via the v2 client, then remove it via the v1
client.  Workers could be operating on one or both versions of the queue while
migration is going on, depending on your needs.

=cut

# vim: ts=4 sts=4 sw=4 et:


lib/MongoDBx/Queue/V1.pm
use 5.010;
use strict;
use warnings;

package MongoDBx::Queue::V1;

# V1 implementation

our $VERSION = '2.001';

use Moose 2;

use Tie::IxHash;
use boolean;
use namespace::autoclean;

my $ID       = '_id';
my $RESERVED = '_r';
my $PRIORITY = '_p';

with (
    'MooseX::Role::Logger',
    'MooseX::Role::MongoDB' => { -version => 1.000 },
    'MongoDBx::Queue::Role::CommonOptions',
);

sub _build__mongo_default_database { $_[0]->database_name }

sub _build__mongo_client_options {
    return {
        write_concern => { w     => "majority" },
        read_concern  => { level => "majority" },
        %{ $_[0]->client_options },
    };
}

sub create_indexes {
    my ($self) = @_;
    # ensure index on PRIORITY in the same order we use for reserving
    $self->_mongo_collection( $self->collection_name )
      ->indexes->create_one( [ $PRIORITY => 1 ] );
}

sub add_task {
    my ( $self, $data, $opts ) = @_;

    $self->_mongo_collection( $self->collection_name )
      ->insert_one( { %$data, $PRIORITY => $opts->{priority} // time() } );
}

sub reserve_task {
    my ( $self, $opts ) = @_;

    my $now = time();
    return $self->_mongo_collection( $self->collection_name )->find_one_and_update(
        {
            $PRIORITY => { '$lte'    => $opts->{max_priority} // $now },
            $RESERVED => { '$exists' => boolean::false },
        },
        { '$set' => { $RESERVED => $now } },
        { sort   => [ $PRIORITY => 1 ] },
    );
}

sub reschedule_task {
    my ( $self, $task, $opts ) = @_;
    $self->_mongo_collection( $self->collection_name )->update_one(
        { $ID => $task->{$ID} },
        {
            '$unset' => { $RESERVED => 0 },
            '$set'   => { $PRIORITY => $opts->{priority} // $task->{$PRIORITY} },
        },
    );
}

sub remove_task {
    my ( $self, $task ) = @_;
    $self->_mongo_collection( $self->collection_name )
      ->delete_one( { $ID => $task->{$ID} } );
}

sub apply_timeout {
    my ( $self, $timeout ) = @_;
    $timeout //= 120;
    my $cutoff = time() - $timeout;
    $self->_mongo_collection( $self->collection_name )->update_many(
        { $RESERVED => { '$lt'     => $cutoff } },
        { '$unset'  => { $RESERVED => 0 } },
    );
}

sub search {
    my ( $self, $query, $opts ) = @_;
    $query = {} unless ref $query eq 'HASH';
    $opts  = {} unless ref $opts eq 'HASH';
    if ( exists $opts->{reserved} ) {
        $query->{$RESERVED} =
          { '$exists' => $opts->{reserved} ? boolean::true : boolean::false };
        delete $opts->{reserved};
    }
    my $cursor =
      $self->_mongo_collection( $self->collection_name )->find( $query, $opts );
    return $cursor->all;
}

sub peek {
    my ( $self, $task ) = @_;
    my @result = $self->search( { $ID => $task->{$ID} } );
    return wantarray ? @result : $result[0];
}

sub size {
    my ($self) = @_;
    return $self->_mongo_collection( $self->collection_name )->count_documents( {} );
}

sub waiting {
    my ($self) = @_;
    return $self->_mongo_collection( $self->collection_name )
      ->count_documents( { $RESERVED => { '$exists' => boolean::false } } );
}

__PACKAGE__->meta->make_immutable;

1;

# vim: ts=4 sts=4 sw=4 et:


lib/MongoDBx/Queue/V2.pm
use 5.010;
use strict;
use warnings;

package MongoDBx::Queue::V2;

# V2 implementation

our $VERSION = '2.001';

use Moose 2;

use Tie::IxHash;
use boolean;
use namespace::autoclean;

my $ID       = '_id';
my $RESERVED = '_x';
my $RES_TIME = '_r';
my $PRIORITY = '_p';

with(
    'MooseX::Role::Logger',
    'MooseX::Role::MongoDB' => { -version => 1.000 },
    'MongoDBx::Queue::Role::CommonOptions',
);

sub _build__mongo_default_database { $_[0]->database_name }

sub _build__mongo_client_options {
    return {
        write_concern => { w     => "majority" },
        read_concern  => { level => "majority" },
        %{ $_[0]->client_options },
    };
}

sub create_indexes {
    my ($self) = @_;
    # ensure index on PRIORITY in the same order we use for reserving
    $self->_mongo_collection( $self->collection_name )
      ->indexes->create_one( [ $RESERVED => 1, $PRIORITY => 1 ] );
    $self->_mongo_collection( $self->collection_name )
      ->indexes->create_one( [ $RESERVED => 1, $RES_TIME => 1 ] );
}

sub add_task {
    my ( $self, $data, $opts ) = @_;

    delete $data->{$RES_TIME};
    $self->_mongo_collection( $self->collection_name )->insert_one(
        {
            %$data,
            $RESERVED => boolean::false,
            $PRIORITY => $opts->{priority} // time(),
        }
    );
}

sub reserve_task {
    my ( $self, $opts ) = @_;

    my $now = time();
    return $self->_mongo_collection( $self->collection_name )->find_one_and_update(
        {
            $RESERVED => boolean::false,
            $PRIORITY => { '$lte' => $opts->{max_priority} // $now },
        },
        {
            '$set' => {
                $RESERVED => boolean::true,
                $RES_TIME => $now,
            }
        },
        { sort => [ $PRIORITY => 1 ] },
    );
}

sub reschedule_task {
    my ( $self, $task, $opts ) = @_;
    $self->_mongo_collection( $self->collection_name )->update_one(
        { $ID => $task->{$ID} },
        {
            '$unset' => { $RES_TIME => 1 },
            '$set'   => {
                $RESERVED => boolean::false,
                $PRIORITY => $opts->{priority} // $task->{$PRIORITY}
            },
        },
    );
}

sub remove_task {
    my ( $self, $task ) = @_;
    $self->_mongo_collection( $self->collection_name )
      ->delete_one( { $ID => $task->{$ID} } );
}

sub apply_timeout {
    my ( $self, $timeout ) = @_;
    $timeout //= 120;
    my $cutoff = time() - $timeout;
    $self->_mongo_collection( $self->collection_name )->update_many(
        { $RESERVED => boolean::true, $RES_TIME => { '$lt' => $cutoff } },
        {
            '$unset' => { $RES_TIME => 1 },
            '$set'   => { $RESERVED => boolean::false },
        },
    );
}

sub search {
    my ( $self, $query, $opts ) = @_;
    $query = {} unless ref $query eq 'HASH';
    $opts  = {} unless ref $opts eq 'HASH';
    if ( exists $opts->{reserved} ) {
        $query->{$RESERVED} = ( $opts->{reserved} ? boolean::true : boolean::false );
        delete $opts->{reserved};
    }
    my $cursor =
      $self->_mongo_collection( $self->collection_name )->find( $query, $opts );
    return $cursor->all;
}

sub peek {
    my ( $self, $task ) = @_;
    my @result = $self->search( { $ID => $task->{$ID} } );
    return wantarray ? @result : $result[0];
}

sub size {
    my ($self) = @_;
    return $self->_mongo_collection( $self->collection_name )->count_documents( {} );
}

sub waiting {
    my ($self) = @_;
    return $self->_mongo_collection( $self->collection_name )
      ->count_documents( { $RESERVED => boolean::false } );
}

__PACKAGE__->meta->make_immutable;

1;

# vim: ts=4 sts=4 sw=4 et:


lib/MongoDBx/Queue/_V1.pm
use 5.010;
use strict;
use warnings;

package MongoDBx::Queue::_V1;

# V1 implementation

our $VERSION = '2.003';

use Moose 2;

use Tie::IxHash;
use boolean;
use namespace::autoclean;

my $ID       = '_id';
my $RESERVED = '_r';
my $PRIORITY = '_p';

with (
    'MooseX::Role::Logger',
    'MooseX::Role::MongoDB' => { -version => 1.000 },
    'MongoDBx::Queue::Role::_CommonOptions',
);

sub _build__mongo_default_database { $_[0]->database_name }

sub _build__mongo_client_options {
    return {
        write_concern => { w     => "majority" },
        read_concern  => { level => "majority" },
        %{ $_[0]->client_options },
    };
}

sub create_indexes {
    my ($self) = @_;
    # ensure index on PRIORITY in the same order we use for reserving
    $self->_mongo_collection( $self->collection_name )
      ->indexes->create_one( [ $PRIORITY => 1 ] );
}

sub add_task {
    my ( $self, $data, $opts ) = @_;

    $self->_mongo_collection( $self->collection_name )
      ->insert_one( { %$data, $PRIORITY => $opts->{priority} // time() } );
}

sub reserve_task {
    my ( $self, $opts ) = @_;

    my $now = time();
    return $self->_mongo_collection( $self->collection_name )->find_one_and_update(
        {
            $PRIORITY => { '$lte'    => $opts->{max_priority} // $now },
            $RESERVED => { '$exists' => boolean::false },
        },
        { '$set' => { $RESERVED => $now } },
        { sort   => [ $PRIORITY => 1 ] },
    );
}

sub reschedule_task {
    my ( $self, $task, $opts ) = @_;
    $self->_mongo_collection( $self->collection_name )->update_one(
        { $ID => $task->{$ID} },
        {
            '$unset' => { $RESERVED => 0 },
            '$set'   => { $PRIORITY => $opts->{priority} // $task->{$PRIORITY} },
        },
    );
}

sub remove_task {
    my ( $self, $task ) = @_;
    $self->_mongo_collection( $self->collection_name )
      ->delete_one( { $ID => $task->{$ID} } );
}

sub apply_timeout {
    my ( $self, $timeout ) = @_;
    $timeout //= 120;
    my $cutoff = time() - $timeout;
    $self->_mongo_collection( $self->collection_name )->update_many(
        { $RESERVED => { '$lt'     => $cutoff } },
        { '$unset'  => { $RESERVED => 0 } },
    );
}

sub search {
    my ( $self, $query, $opts ) = @_;
    $query = {} unless ref $query eq 'HASH';
    $opts  = {} unless ref $opts eq 'HASH';
    if ( exists $opts->{reserved} ) {
        $query->{$RESERVED} =
          { '$exists' => $opts->{reserved} ? boolean::true : boolean::false };
        delete $opts->{reserved};
    }
    my $cursor =
      $self->_mongo_collection( $self->collection_name )->find( $query, $opts );
    return $cursor->all;
}

sub peek {
    my ( $self, $task ) = @_;
    my @result = $self->search( { $ID => $task->{$ID} } );
    return wantarray ? @result : $result[0];
}

sub size {
    my ($self) = @_;
    return $self->_mongo_collection( $self->collection_name )->estimated_document_count();
}

sub waiting {
    my ($self) = @_;
    return $self->_mongo_collection( $self->collection_name )
      ->count_documents( { $RESERVED => { '$exists' => boolean::false } } );
}

__PACKAGE__->meta->make_immutable;

1;

# vim: ts=4 sts=4 sw=4 et:


lib/MongoDBx/Queue/_V2.pm
use 5.010;
use strict;
use warnings;

package MongoDBx::Queue::_V2;

# V2 implementation

our $VERSION = '2.003';

use Moose 2;

use Tie::IxHash;
use boolean;
use namespace::autoclean;

my $ID       = '_id';
my $RESERVED = '_x';
my $RES_TIME = '_r';
my $PRIORITY = '_p';

with(
    'MooseX::Role::Logger',
    'MooseX::Role::MongoDB' => { -version => 1.000 },
    'MongoDBx::Queue::Role::_CommonOptions',
);

sub _build__mongo_default_database { $_[0]->database_name }

sub _build__mongo_client_options {
    return {
        write_concern => { w     => "majority" },
        read_concern  => { level => "majority" },
        %{ $_[0]->client_options },
    };
}

sub create_indexes {
    my ($self) = @_;
    # ensure index on PRIORITY in the same order we use for reserving
    $self->_mongo_collection( $self->collection_name )
      ->indexes->create_one( [ $RESERVED => 1, $PRIORITY => 1 ] );
    $self->_mongo_collection( $self->collection_name )
      ->indexes->create_one( [ $RESERVED => 1, $RES_TIME => 1 ] );
}

sub add_task {
    my ( $self, $data, $opts ) = @_;

    delete $data->{$RES_TIME};
    $self->_mongo_collection( $self->collection_name )->insert_one(
        {
            %$data,
            $RESERVED => boolean::false,
            $PRIORITY => $opts->{priority} // time(),
        }
    );
}

sub reserve_task {
    my ( $self, $opts ) = @_;

    my $now = time();
    return $self->_mongo_collection( $self->collection_name )->find_one_and_update(
        {
            $RESERVED => boolean::false,
            $PRIORITY => { '$lte' => $opts->{max_priority} // $now },
        },
        {
            '$set' => {
                $RESERVED => boolean::true,
                $RES_TIME => $now,
            }
        },
        { sort => [ $PRIORITY => 1 ] },
    );
}

sub reschedule_task {
    my ( $self, $task, $opts ) = @_;
    $self->_mongo_collection( $self->collection_name )->update_one(
        { $ID => $task->{$ID} },
        {
            '$unset' => { $RES_TIME => 1 },
            '$set'   => {
                $RESERVED => boolean::false,
                $PRIORITY => $opts->{priority} // $task->{$PRIORITY}
            },
        },
    );
}

sub remove_task {
    my ( $self, $task ) = @_;
    $self->_mongo_collection( $self->collection_name )
      ->delete_one( { $ID => $task->{$ID} } );
}

sub apply_timeout {
    my ( $self, $timeout ) = @_;
    $timeout //= 120;
    my $cutoff = time() - $timeout;
    $self->_mongo_collection( $self->collection_name )->update_many(
        { $RESERVED => boolean::true, $RES_TIME => { '$lt' => $cutoff } },
        {
            '$unset' => { $RES_TIME => 1 },
            '$set'   => { $RESERVED => boolean::false },
        },
    );
}

sub search {
    my ( $self, $query, $opts ) = @_;
    $query = {} unless ref $query eq 'HASH';
    $opts  = {} unless ref $opts eq 'HASH';
    if ( exists $opts->{reserved} ) {
        $query->{$RESERVED} = ( $opts->{reserved} ? boolean::true : boolean::false );
        delete $opts->{reserved};
    }
    my $cursor =
      $self->_mongo_collection( $self->collection_name )->find( $query, $opts );
    return $cursor->all;
}

sub peek {
    my ( $self, $task ) = @_;
    my @result = $self->search( { $ID => $task->{$ID} } );
    return wantarray ? @result : $result[0];
}

sub size {
    my ($self) = @_;
    return $self->_mongo_collection( $self->collection_name )->estimated_document_count();
}

sub waiting {
    my ($self) = @_;
    return $self->_mongo_collection( $self->collection_name )
      ->count_documents( { $RESERVED => boolean::false } );
}

__PACKAGE__->meta->make_immutable;

1;

# vim: ts=4 sts=4 sw=4 et:


Makefile.PL
use strict;
use warnings;

use 5.010;

use ExtUtils::MakeMaker 6.17;

my %WriteMakefileArgs = (
  "ABSTRACT" => "A message queue implemented with MongoDB",
  "AUTHOR" => "David Golden <dagolden\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.17"
  },
  "DISTNAME" => "MongoDBx-Queue",
  "LICENSE" => "apache",
  "MIN_PERL_VERSION" => "5.010",
  "NAME" => "MongoDBx::Queue",
  "PREREQ_PM" => {
    "MongoDB" => 2,
    "Moose" => 2,
    "Moose::Role" => 2,
    "MooseX::AttributeShortcuts" => 0,
    "MooseX::Role::Logger" => 0,
    "MooseX::Role::MongoDB" => 1,
    "MooseX::Types::Moose" => 0,
    "Tie::IxHash" => 0,
    "boolean" => 0,
    "namespace::autoclean" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::Deep" => 0,
    "Test::More" => "0.96"
  },
  "VERSION" => "2.003",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "MongoDB" => 2,
  "Moose" => 2,
  "Moose::Role" => 2,
  "MooseX::AttributeShortcuts" => 0,
  "MooseX::Role::Logger" => 0,
  "MooseX::Role::MongoDB" => 1,
  "MooseX::Types::Moose" => 0,
  "Test::Deep" => 0,
  "Test::More" => "0.96",
  "Tie::IxHash" => 0,
  "boolean" => 0,
  "namespace::autoclean" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
